### PR TITLE
Olofproperdtlsfix

### PR DESCRIPTION
**Pull request by olofkallander**
several issues with reduced mem fottprint and udpendpoint
Mostly visible when running dtlsRace tests.

### CODE REVIEW

File: test/transport/IceIntegrationTest.cpp [hunk 1]
```diff
@@ -293,7 +293,7 @@ TEST_F(IceIntegrationTest, dtlsRace)
     }
 
     bool connected = false;
-    while (!connected && utils::Time::getAbsoluteTime() - startTime < 5 * utils::Time::sec)
+    while (!connected && utils::Time::getAbsoluteTime() - startTime < 20 * utils::Time::sec)
```
Comment (olofkallander) [inline test/transport/IceIntegrationTest.cpp:296]: will revert before final commit

File: transport/ice/IceSession.cpp [hunk 2]
```diff
@@ -525,19 +525,6 @@ void IceSession::onResponseReceived(IceEndpoint* endpoint,
     {
         return;
     }
```
Comment (olofkallander) [inline transport/ice/IceSession.cpp:527]: responses should not really contain USERNAME so we should not waste time on it.

File: transport/ice/IceSession.cpp [hunk 3]
```diff
@@ -694,11 +687,6 @@ void IceSession::sendResponse(IceEndpoint* endpoint,
     {
         response.add(StunError(code, errorPhrase));
     }
```
Comment (olofkallander) [inline transport/ice/IceSession.cpp:689]: We should not add USERNAME either to responses.

File: transport/BaseUdpEndpoint.cpp [hunk 4]
```diff
@@ -66,8 +39,8 @@ BaseUdpEndpoint::BaseUdpEndpoint(const char* name,
     : _state(Endpoint::CLOSED),
       _name(name),
       _localPort(localPort),
-      _receiveJobs(jobManager, 16),
-      _sendJobs(jobManager, 16),
+      _receiveJobs(jobManager, maxSessionCount),
+      _sendJobs(jobManager, maxSessionCount),
```
Comment (olofkallander) [inline transport/BaseUdpEndpoint.cpp:43]: Will rever to 16.

File: transport/BaseUdpEndpoint.cpp [hunk 5]
```diff
@@ -66,8 +39,8 @@ BaseUdpEndpoint::BaseUdpEndpoint(const char* name,
     : _state(Endpoint::CLOSED),
       _name(name),
       _localPort(localPort),
-      _receiveJobs(jobManager, 16),
-      _sendJobs(jobManager, 16),
+      _receiveJobs(jobManager, maxSessionCount),
```
Comment (olofkallander) [inline transport/BaseUdpEndpoint.cpp:42]: A small queue here is fine for receive jobs as they are like 2, but the queue is also used for unreg jobs and that can fill the queue causing receive to halt forever.

File: transport/TransportImpl.cpp [hunk 6]
```diff
@@ -1065,8 +1065,10 @@ void TransportImpl::internalIceReceived(Endpoint& endpoint,
             return;
         }
 
-        if (_rtpIceSession->getState() == ice::IceSession::State::CONNECTING)
+        //logger::debug("ice received, ICE state %u", _loggableId.c_str(), _rtpIceSession->getState());
+        if (_rtpIceSession->getState() != ice::IceSession::State::CONNECTED)
```
Comment (olofkallander) [inline transport/TransportImpl.cpp:1069]: Not usre about this yet. Something goes wron un unreg so there is a ref count 1 on transportimpl.

File: transport/UdpEndpointImpl.cpp [hunk 7]
```diff
@@ -75,7 +39,7 @@ void UdpEndpointImpl::sendStunTo(const transport::SocketAddress& target,
     uint64_t timestamp)
 {
     auto* msg = ice::StunMessage::fromPtr(data);
-    if (msg->header.isRequest() && !_dtlsListeners.contains(target) && !_iceResponseListeners.contains(transactionId))
+    if (msg->header.isRequest() && !_iceResponseListeners.contains(transactionId))
```
Comment (olofkallander) [inline transport/UdpEndpointImpl.cpp:42]: I turned to relying only on transaction ids for response routing. It is the right way.

File: transport/TransportImpl.cpp [hunk 8]
```diff
@@ -1065,7 +1065,8 @@ void TransportImpl::internalIceReceived(Endpoint& endpoint,
             return;
         }
 
-        if (_rtpIceSession->getState() == ice::IceSession::State::CONNECTING)
+        const auto iceState = _rtpIceSession->getState();
+        if (iceState == ice::IceSession::State::CONNECTING || iceState == ice::IceSession::State::READY)
         {
             endpoint.registerListener(source, this);
```
Comment (olofkallander) [inline transport/TransportImpl.cpp:1071]: What happened was that a packet was inbound on the udp endpoint and fit the user name.  Ice session was already in idle and transport just unregistered. But the packet slipped in and caused registration and packet to go to idle ice session.

File: transport/UdpEndpointImpl.cpp [hunk 9]
```diff
@@ -93,42 +64,43 @@ void UdpEndpointImpl::sendStunTo(const transport::SocketAddress& target,
...
-    logger::debug("unregister %p", _name.c_str(), listener);
+    LOG("unregister %p", _name.c_str(), listener);
     for (auto& item : _iceListeners)
     {
         if (item.second == listener)
         {
             _iceListeners.erase(item.first);
-            _receiveJobs.addJob<UnRegisterNotifyListenerJob>(*this, *listener);
+            listener->onUnregistered(*this);
+            break;
```
Comment (olofkallander) [inline transport/UdpEndpointImpl.cpp:103]: only one username is registered per transport

File: transport/UdpEndpointImpl.cpp [hunk 10]
```diff
@@ -252,32 +239,61 @@ void UdpEndpointImpl::registerListener(const std::string& stunUserName, IEvents*
...
+        return;
     }
+
+    logger::warn("dtls listener swap on %s skipped. Already removed", _name.c_str(), srcAddress.toString().c_str());
 }
 
 void UdpEndpointImpl::focusListener(const SocketAddress& remotePort, IEvents* listener)
 {
-    return;
+    _receiveJobs.post([this, remotePort, listener]() {
```
Comment (olofkallander) [inline transport/UdpEndpointImpl.cpp:283]: I do not understand how skipping this code was helping. focusListener was not called in the failing test dtlsRace.

File: test/transport/IceIntegrationTest.cpp (hunk 1 above)
Comment (olofkallander) [reply test/transport/IceIntegrationTest.cpp]: It will exit as soon as they are connected. Giving them a bit more time since there are many connections.

**Review by RicardoMDomingues — APPROVED** [review]

**Pull request merged**